File: process_data.R
# Get data from database ----
library(RPostgreSQL)
pg <- dbConnect(PostgreSQL())

ct_orders <- dbGetQuery(pg, "SELECT file_name, order_text FROM ct_orders.raw_text")

rs <- dbDisconnect(pg)

regex <- "^.*excluded.*?(10-Q|10-K|Form.*?)\\sfiled\\s+(?:on\\s)?(.*?\\d{4}).*"
ct_orders$details <- gsub(regex, "\\1;\\2", gsub("\n", " ", ct_orders$order_text), perl=TRUE)
ct_orders$form <- unlist(lapply(strsplit(ct_orders$details, ";"), function(x) x[1]))
ct_orders$form_date <- unlist(lapply(strsplit(ct_orders$details, ";"), function(x) x[2]))
ct_orders$details <- NULL
ct_orders$form_date <- as.Date(ct_orders$form_date, "%B %d, %Y")
table(is.na(ct_orders$form_date))

# ct_orders$order_type <- gsub("^.*ORDER\\s+(\\w+).*$", "\\1", ct_orders$order_text)
ct_orders$denied <- grepl("ORDER\\s+DENYING", ct_orders$order_text)
ct_orders$granted <- grepl("ORDER\\s+GRANTING", ct_orders$order_text)

# Get exhibits mentioned
matches <- gregexpr("[Ee]xhibit\\s+[0-9][A-Z0-9a-z\\(\\)\\.-]+", ct_orders$order_text)
matched_text <- regmatches(ct_orders$order_text , matches)
ct_orders$exhibits <- unlist(lapply(matched_text, function(x) paste(x, collapse=";")))
ct_orders$order_text <- NULL

# Put data into my database ----
library(RPostgreSQL)
pg <- dbConnect(PostgreSQL())

dbWriteTable(pg, c("ct_orders", "processed_data"), ct_orders,
             overwrite=TRUE, row.names=FALSE)

dbGetQuery(pg, "ALTER TABLE ct_orders.processed_data
                    ALTER COLUMN exhibits TYPE text[]
                        USING regexp_split_to_array(exhibits, ';')")

dbGetQuery(pg, "ALTER TABLE ct_orders.processed_data OWNER TO ct_orders_access")

rs <- dbDisconnect(pg)

File: process_data.py

# coding: utf-8

# Get raw data from database

import pandas as pd

# Get data from database
import sqlalchemy as sa
import psycopg2 as pg
from pandas.io.sql import read_sql

from sqlalchemy import create_engine
engine = create_engine('postgresql://iangow.me/crsp')

sql = "SELECT file_name, order_text FROM ct_orders.raw_text"

df = pd.read_sql(sa.text(sql), engine)


# Extract details of affected form type and date

def extract_details(order_text):
    import re
    regex = r"^.*excluded.*?(10-Q|10-K|Form.*?)\sfiled\s+(?:on\s)?(.*?\d{4}).*"
    # regex = r"excluded"
    matches = re.search(regex, order_text, flags=re.S)
    if matches:
        return matches

df['details'] = df['order_text'].apply(extract_details)

def form_type(detail_text):
    import re
    if detail_text:
        temp = detail_text.group(1)
        return re.sub(r"\n", " ", temp)

## Extract form type
df['form'] = df['details'].apply(form_type)

def convert_date(datetext):
    from datetime import datetime
    import re

    if datetext:
        try:
            datetext = re.sub(r"\n", " ", datetext)
            date = datetime.strptime(datetext, '%B %d, %Y')
            return date
        except:
            return None

# Extract form date
def form_date(detail_text):
    import re
    if detail_text:
        temp = detail_text.group(2)
        return convert_date(temp)

df['form_date'] = df['details'].apply(form_date)


# Get information on whether order grants or denies request

def check_regex(text, pattern):
    import re

    if re.findall(pattern, text, flags=re.S):
        return True
    else:
        return False

# df['order_type'] = df['order_text'].map(lambda x: re.findall(r"^.*ORDER\s+(\w+).*$", x, flags=re.S))
df['denied'] = df['order_text'].map(lambda x: check_regex(x, pattern=r"ORDER\s+DENYING"))
df['granted'] = df['order_text'].map(lambda x: check_regex(x, pattern=r"ORDER\s+GRANTING"))

# Get exhibits mentioned
def extract_exhibits(order_text):
    import re
    regex = r"[Ee]xhibit\s+[0-9][A-Z0-9a-z\(\)\.-]+"
    matches = re.findall(regex, order_text, flags=re.S)
    if matches:
        return matches

df['exhibits'] = df['order_text'].apply(extract_exhibits)

# Code to push data to PostgreSQL database

# Delete columns not needed any more
del df['order_text']
del df['details']

# Actually push data to PostgreSQL database
df.to_sql('processed_data_py', engine, schema="ct_orders",
         if_exists="replace", index=False)

# Do some database-related clean-up
res = engine.execute(
    """ALTER TABLE ct_orders.processed_data_py
    OWNER TO ct_orders_access""")
